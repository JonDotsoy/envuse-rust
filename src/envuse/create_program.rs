use crate::errors::program_error::ProgramError;

use crate::syntax_error::SyntaxError;

use super::super::parser::ast::AST;
use super::super::parser::tokenizer::Tokenizer;
use super::program::Program;

pub fn create_program<T: ToString>(
    source: T,
    location: Option<T>,
) -> Result<Program, Box<dyn std::error::Error>> {
    let location_val = location.map(|t| t.to_string());

    let result_program: Result<Program, Box<dyn std::error::Error>> = try {
        Program {
            location: location_val.clone(),
            source: source.to_string(),
            ast: AST::parse(Tokenizer::parse(source.to_string())?)?,
        }
    };

    match result_program {
        Err(error) if error.is::<SyntaxError>() => {
            let syntax_error = error.downcast::<SyntaxError>().unwrap();
            do yeet ProgramError {
                message: format!("SyntaxError: {}", syntax_error.message),
                span: Some(syntax_error.span.clone()),
                source: source.to_string(),
                location: location_val,
                cause: Some(*syntax_error),
            }
        }
        result_program => result_program,
    }
}
