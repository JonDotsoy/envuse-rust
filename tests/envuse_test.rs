#![feature(try_blocks)]

#[cfg(test)]
mod envuse_test {
    use std::collections::BTreeMap;

    use envuse_parser::create_program;
    use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot};

    #[test]
    fn should_catch_program_error() {
        let program = create_program(r#"FOO "biz""#, None).err().unwrap();

        // dbg!(program);

        assert_debug_snapshot!(program);
        assert_snapshot!(program.to_string());
    }

    #[test]
    fn should_catch_program_error_to_js() {
        let error = create_program("\n\n\tFOO 123", Some("/app/.envuse"))
            .err()
            .unwrap();
        assert_snapshot!(error.to_string());

        let error = create_program("\n\n\tFOO 123", None).err().unwrap();
        assert_snapshot!(error.to_string());
    }

    #[test]
    fn should_create_program() {
        let program = create_program(r#"FOO="biz""#, None).unwrap();

        assert_debug_snapshot!(program);
    }

    #[test]
    fn should_use_of_parse_must_not_break_with_different_syntax() {
        let program = create_program(r#"FOO="biz""#, None).unwrap();

        program
            .parse(
                BTreeMap::from([
                    (String::from("AAA"), None),
                    (String::from("FOO"), Some(String::from(""))),
                    (String::from("JUM"), None),
                    (String::from("TAZ"), None),
                ]),
                None,
            )
            .unwrap();

        program
            .parse([("AAA", ""), ("FOO", ""), ("JUM", ""), ("TAZ", "")], None)
            .unwrap();

        program
            .parse(
                [
                    ("AAA", None),
                    ("FOO", Some("")),
                    ("JUM", None),
                    ("TAZ", None),
                ],
                None,
            )
            .unwrap();
    }

    #[test]
    fn should_parse_an_string_value() {
        let program = create_program(r#"FOO="biz""#, None).unwrap();

        assert_debug_snapshot!(program.parse([("FOO", "BAR")], None));
    }

    #[test]
    fn should_parse_an_number_value() {
        let program = create_program(r#"FOO:Number"#, None).unwrap();

        assert_debug_snapshot!(program.parse([("FOO", "30_000")], None));
    }

    #[test]
    fn should_parse_an_boolean_value() {
        let program = create_program(r#"FOO:Boolean"#, None).unwrap();

        assert_debug_snapshot!(program.parse([("FOO", "true")], None));
    }

    #[test]
    fn should_parse_an_null_value() {
        let program = create_program(r#"FOO: String?"#, None).unwrap();

        assert_debug_snapshot!(program.parse([("BAR", "true")], None));
    }

    #[test]
    fn should_read_the_default_values() {
        let program = create_program(r#"FOO: String = "Biz""#, None)
            .map_err(|e| panic!("{}", e.to_string()))
            .unwrap();

        assert_debug_snapshot!(program.parse([("BAR", "true")], None));
    }

    #[test]
    fn should_read_the_default_values_2() {
        let program = create_program(r#"FOO = "Biz""#, None)
            .map_err(|e| panic!("{}", e.to_string()))
            .unwrap();

        assert_debug_snapshot!(program.parse([("BAR", "true")], None));
    }

    #[test]
    fn should_full_sample() {
        let source = r#"
            FOO: String = "val"
            BAR: Number = 1223
            BIZ: Boolean = "true"
            BLI: String?
        "#;

        let program = create_program(source, Some(".envuse"))
            .map_err(|e| panic!("{}", e.to_string()))
            .unwrap();

        let envs: [(&str, &str); 0] = [];

        assert_yaml_snapshot!(program.parse(envs, None).unwrap());
    }

    #[test]
    fn should_catch_error_type_unknown() {
        let source = r###"
            FOO: unknown
        "###;

        let program = create_program(source, Some(".envuse")).unwrap();
        let error = program.parse(None, None).unwrap_err();

        assert_snapshot!(error.to_string());
        assert_debug_snapshot!(error);
    }

    #[test]
    fn should_parse_custom_types() {
        let source = r###"
            FOO: FIZ
        "###;

        let program = create_program(source, Some(".envuse")).unwrap();

        let custom_transformers = ["FIZ"];
        let values = [("FOO", "bar")];

        let parsed = program.parse(values, custom_transformers).unwrap();

        assert_debug_snapshot!(parsed);
    }
}
