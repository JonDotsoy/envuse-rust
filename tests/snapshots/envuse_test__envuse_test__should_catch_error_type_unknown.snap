---
source: tests/envuse_test.rs
expression: error.to_string()
---
ParseError: Type unknown is not valid type

>    2 |             FOO: unknown
                     ▀▀▀▀▀▀▀▀▀▀▀▀
    at .envuse:2:13

